File: php/steps/step04_hello_agent.php
#!/usr/bin/env php
<?php
/**My first AI phone agent -- Hello World edition.*/

require __DIR__ . '/../../sdks/signalwire-php/vendor/autoload.php';

use SignalWire\Agent\AgentBase;
use SignalWire\SWAIG\FunctionResult;

// --- Load .env file ---
$envFile = __DIR__ . '/../.env';
if (file_exists($envFile)) {
    foreach (file($envFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES) as $line) {
        if (str_starts_with(trim($line), '#')) continue;
        putenv(trim($line));
        [$key, $val] = explode('=', $line, 2);
        $_ENV[trim($key)] = trim($val);
    }
}

// --- Auto-detect ngrok tunnel ---
function checkNgrok(): string {
    try {
        $ctx = stream_context_create(['http' => ['timeout' => 1]]);
        $json = @file_get_contents('http://127.0.0.1:4040/api/tunnels', false, $ctx);
        if ($json) {
            $data = json_decode($json, true);
            foreach (($data['tunnels'] ?? []) as $t) {
                if (($t['proto'] ?? '') === 'https') {
                    $url = $t['public_url'];
                    putenv("SWML_PROXY_URL_BASE=$url");
                    $_ENV['SWML_PROXY_URL_BASE'] = $url;
                    echo "ngrok detected: $url\n";
                    $user = getenv('SWML_BASIC_AUTH_USER') ?: '';
                    $pw   = getenv('SWML_BASIC_AUTH_PASSWORD') ?: '';
                    if ($user && $pw) {
                        $parsed = parse_url($url);
                        $full = "{$parsed['scheme']}://$user:$pw@{$parsed['host']}/";
                        echo "\n  SignalWire SWML URL (paste into dashboard):\n  $full\n\n";
                    }
                    return $url;
                }
            }
        }
    } catch (\Exception $e) {}
    $current = getenv('SWML_PROXY_URL_BASE') ?: '';
    if ($current) {
        echo "Using SWML_PROXY_URL_BASE from .env: $current\n";
    } else {
        echo "No ngrok tunnel detected and SWML_PROXY_URL_BASE not set\n";
    }
    return $current;
}

checkNgrok();

// --- Agent ---

$agent = new AgentBase(name: 'hello-agent', route: '/');

$agent->addLanguage(name: 'English', code: 'en-US', voice: 'rime.spore');

$agent->promptAddSection(
    'Role',
    'You are a friendly assistant named Buddy. '
    . 'You greet callers warmly, ask how their day is going, '
    . 'and have a brief pleasant conversation. '
    . 'Keep your responses short since this is a phone call.'
);

$agent->setPostPrompt(
    'Summarize this conversation in 2-3 sentences. '
    . 'Include what the caller wanted and how the conversation went.'
);

$agent->onSummary(function ($summary, $rawData) {
    @mkdir('calls', 0755, true);
    $callId = $rawData['call_id'] ?? date('Ymd_His');
    $path = "calls/$callId.json";
    file_put_contents($path, json_encode($rawData, JSON_PRETTY_PRINT));
    echo "Call summary saved: $path\n";
});

$agent->run();

File: dotnet/steps/step04_hello_agent.cs
// My first AI phone agent -- Hello World edition.

using SignalWire.Agent;
using SignalWire.SWAIG;
using System.Text.Json;

// --- Load .env file ---
var envFile = Path.Combine(AppContext.BaseDirectory, "../../../.env");
if (!File.Exists(envFile)) envFile = Path.Combine(Directory.GetCurrentDirectory(), ".env");
if (File.Exists(envFile))
{
    foreach (var line in File.ReadAllLines(envFile))
    {
        var trimmed = line.Trim();
        if (string.IsNullOrEmpty(trimmed) || trimmed.StartsWith('#')) continue;
        var idx = trimmed.IndexOf('=');
        if (idx > 0)
            Environment.SetEnvironmentVariable(trimmed[..idx].Trim(), trimmed[(idx + 1)..].Trim());
    }
}

// --- Auto-detect ngrok tunnel ---
try
{
    using var http = new HttpClient { Timeout = TimeSpan.FromSeconds(1) };
    var json = await http.GetStringAsync("http://127.0.0.1:4040/api/tunnels");
    var doc = JsonDocument.Parse(json);
    foreach (var tunnel in doc.RootElement.GetProperty("tunnels").EnumerateArray())
    {
        if (tunnel.GetProperty("proto").GetString() == "https")
        {
            var url = tunnel.GetProperty("public_url").GetString()!;
            Environment.SetEnvironmentVariable("SWML_PROXY_URL_BASE", url);
            Console.WriteLine($"ngrok detected: {url}");
            var user = Environment.GetEnvironmentVariable("SWML_BASIC_AUTH_USER") ?? "";
            var pw = Environment.GetEnvironmentVariable("SWML_BASIC_AUTH_PASSWORD") ?? "";
            if (user != "" && pw != "")
            {
                var uri = new Uri(url);
                Console.WriteLine($"\n  SignalWire SWML URL:\n  {uri.Scheme}://{user}:{pw}@{uri.Host}/\n");
            }
            break;
        }
    }
}
catch
{
    var current = Environment.GetEnvironmentVariable("SWML_PROXY_URL_BASE") ?? "";
    if (current != "")
        Console.WriteLine($"Using SWML_PROXY_URL_BASE from .env: {current}");
    else
        Console.WriteLine("No ngrok tunnel detected and SWML_PROXY_URL_BASE not set");
}

// --- Agent ---

var agent = new AgentBase(new AgentOptions { Name = "hello-agent" });

agent.AddLanguage("English", "en-US", "rime.spore");

agent.PromptAddSection(
    "Role",
    "You are a friendly assistant named Buddy. "
    + "You greet callers warmly, ask how their day is going, "
    + "and have a brief pleasant conversation. "
    + "Keep your responses short since this is a phone call."
);

agent.SetPostPrompt(
    "Summarize this conversation in 2-3 sentences. "
    + "Include what the caller wanted and how the conversation went."
);

agent.OnSummary((summary, rawData, headers) =>
{
    Directory.CreateDirectory("calls");
    var callId = rawData?.GetValueOrDefault("call_id")?.ToString()
                 ?? DateTime.Now.ToString("yyyyMMdd_HHmmss");
    var path = $"calls/{callId}.json";
    File.WriteAllText(path, JsonSerializer.Serialize(rawData, new JsonSerializerOptions { WriteIndented = true }));
    Console.WriteLine($"Call summary saved: {path}");
});

agent.Run();
